### PR TITLE
Discover the services of the mail server  And prepare the migration

### PR DESCRIPTION
**Pull request by stephdl**
With this PR I bring some changes

the tcp port of submission and imap are found and set before the service start (we query a local redis replica)
we do not offer anymore to use a custom mail server, we test the cluster to find and make an array of all mail server of the cluster to populate a dropdown
the action import-module is created to load the roundcubemail.sql dump from NS7

Capture d’écran du 2022-12-03 15-21-41
Capture d’écran du 2022-12-03 15-20-56
Capture d’écran du 2022-12-03 15-20-45

### CODE REVIEW

File: imageroot/bin/discover-service [hunk 1]
```diff
@@ -14,12 +14,12 @@ import sys
...
+mail_server = os.environ["MAIL_SERVER"]
 rdb = agent.redis_connect(host="127.0.0.1") # full read-only access on every key
 
 # we query about TCP service provided by the mail server
-smtp_port = rdb.hget(f"module/{mail_server_to_use}/srv/tcp/submission", "port") or ""
-imap_port = rdb.hget(f"module/{mail_server_to_use}/srv/tcp/imap", "port") or ""
+smtp_port = rdb.hget(f"module/{mail_server}/srv/tcp/submission", "port") or ""
+imap_port = rdb.hget(f"module/{mail_server}/srv/tcp/imap", "port") or ""
 
 agent.set_env("ROUNDCUBEMAIL_DEFAULT_PORT", imap_port)
```
Comment (DavidePrincipi) [inline imageroot/bin/discover-service:24]: agent.set_env does not work here: it is not an action step!

You should implement something like https://github.com/NethServer/ns8-kickstart/pull/19

- write a file with relevant vars
- load them in the .service unit (or with podman run --env-file)

File: imageroot/actions/import-module/40Migration [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+exec 1>&2 # Redirect any output to the journal (stderr)
+
+# Prepare an initialization script that restores the dump file
+mkdir -vp initdb.d
+mv -v roundcubemail.sql initdb.d
+#do the bash file to restore and exit once done
+cat - >initdb.d/roundcubemail_restore.sh <<'EOS'
+# Read dump file from standard input:
+cd docker-entrypoint-initdb.d
+mysql -uroundcubemail -proundcubemail roundcubemail < roundcubemail.sql
```
Comment (DavidePrincipi) [inline imageroot/actions/import-module/40Migration:18]: The image entrypoint script already loads .sql dump files. Why does DB restore run here?
```suggestion
```

File: imageroot/actions/import-module/40Migration [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+# Print additional information:
+mysql --version
+# The script is sourced, override entrypoint args and exit:
+set -- true
+exit 1
+EOS
+# file must be executable
+chmod a+x initdb.d/roundcubemail_restore.sh
```
Comment (DavidePrincipi) [inline imageroot/actions/import-module/40Migration:26]: The image entrypoint script includes (`source` aka `.`) .sh files if they are not flagged _executable_.

The `set -- true` line should cause the entrypoint exit by itself, after finishing the load of init scripts. By contrary `exit 1` forces the script termination with error, preventing any further operation. This is an unexpected behavior IMO... Why is it arranged like that?

```suggestion
set -- true
EOS
```

File: imageroot/systemd/user/roundcubemail-app.service [hunk 4]
```diff
@@ -16,6 +16,7 @@ Restart=always
 TimeoutStopSec=70
 ExecStartPre=/bin/mkdir -p %S/state/config
 ExecStartPre=/bin/rm -f %t/roundcubemail-app.pid %t/roundcubemail-app.ctr-id
+ExecStartPre=/usr/local/bin/runagent discover-service
```
Comment (DavidePrincipi) [inline imageroot/systemd/user/roundcubemail-app.service:19]: What to do if the discover-service script fails?

1. abort the unit startup
2. add `-` prefix, ignore its exit code and start anyway with the old configuration, provided it was not overwritten

```suggestion
ExecStartPre=-/usr/local/bin/runagent discover-service
```

File: imageroot/actions/import-module/40Migration (hunk 3 above)
Comment (stephdl) [reply imageroot/actions/import-module/40Migration]: the interactive container is not stopped at the end, it is not nice but the exit 1 stops the container

File: imageroot/actions/import-module/40Migration (hunk 2 above)
Comment (stephdl) [reply imageroot/actions/import-module/40Migration]: good idea

File: imageroot/bin/discover-service (hunk 1 above)
Comment (stephdl) [reply imageroot/bin/discover-service]: done

File: imageroot/systemd/user/roundcubemail-app.service (hunk 4 above)
Comment (stephdl) [reply imageroot/systemd/user/roundcubemail-app.service]: done

File: imageroot/actions/import-module/40Migration [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+#
+
+set -e -o pipefail
+exec 1>&2 # Redirect any output to the journal (stderr)
+
+# Prepare an initialization script that restores the dump file
+mkdir -vp initdb.d
+mv -v roundcubemail.sql initdb.d
+#do the bash file to restore and exit once done
+cat - >initdb.d/roundcubemail_restore.sh <<'EOS'
```
[the remark this answers is not in the document]
Comment (stephdl) [reply imageroot/actions/import-module/40Migration]: done

File: ui/public/i18n/language.json [hunk 6]
```diff
@@ -27,32 +27,19 @@
...
-    "encrypt_imap": "IMAP encryption",
-    "none": "None",
-    "starttls": "STARTTLS",
-    "tls": "TLS",
-    "imap_port": "IMAP TCP port",
-    "encrypt_smtp": "SMTP encryption",
-    "smtp_port": "SMTP TCP port",
-    "tls_verify_smtp": "Use a valid certificate for SMTP",
-    "tls_verify_imap": "Use a valid certificate for IMAP",
+    "mail_server_fqdn": "Mail server hostname",
```
Comment (andre8244) [inline ui/public/i18n/language.json:30]:
```suggestion
    "mail_server_fqdn": "Mail server",
```

File: ui/public/i18n/language.json [hunk 7]
```diff
@@ -27,32 +27,19 @@
...
-    "test_smtp": "SMTP error",
-    "cannot_connect_to_server": "Cannot connect to server",
-    "valid_certificate_verification_error": "The certificate is not valid for the hostname or self signed",
     "configuring": "Configuring",
     "instance_configuration": "Configure Roundcube",
     "placeholder_plugins": "Plugins list (comma separated: plugin1,plugin2,plugin3)",
-    "choose": "Choose encryption"
+    "choose_mail_server": "Select a mail server as endpoint",
+    "choose_the_mail_server_to_use": "Roundcube needs to connect to a mail server, select one in the list",
+    "mail_server_is_not_valid": "This mail server cannot be used by roundcube webmail"
```
Comment (andre8244) [inline ui/public/i18n/language.json:42]:
```suggestion
    "mail_server_is_not_valid": "This mail server cannot be used by Roundcube webmail"
```
What is the condition that raises this error? Why a mail server in the list valid should be invalid?

File: imageroot/actions/get-configuration/20read [hunk 8]
```diff
@@ -23,14 +24,27 @@ config["lets_encrypt"] = os.getenv("TRAEFIK_LETS_ENCRYPT") == "True"
...
+config["upload_max_filesize"] = os.getenv("ROUNDCUBEMAIL_UPLOAD_MAX_FILESIZE","10").replace('M','')
+
+rdb = agent.redis_connect(host='127.0.0.1')
+modules=[]
+
+# we query about all mail server
+for key in agent.list_service_providers(rdb,'imap','tcp'):
+
+    # mail_hostname could be optional, use user_domain instead
+    mail_hostname = key['user_domain'] if 'mail_hostname' not in key else key['mail_hostname']
```
Comment (DavidePrincipi) [inline imageroot/actions/get-configuration/20read:36]:
```suggestion
    mail_hostname = key.get('mail_hostname') or key.get('user_domain')
```

Comment (stephdl) [reply imageroot/actions/get-configuration/20read:36]: `mail_hostname = key.get('mail_hostname', key.get('user_domain'))`

does it work :D

Comment (stephdl) [reply imageroot/actions/get-configuration/20read:36]: @davidep, what about to discovery when the mail server hasb been moved, I saw that srv discovery has an event  for this purpose, what do you think ?

**Review by DavidePrincipi — APPROVED** [review]
I'd go by merging this PR and open separate PRs to complete clone-module and restore-module implementation.

**Pull request merged**